### ISSUE TITLE
The DOC phaser

### ISSUE THREAD

**Issue by finanalyst** [Raku/RakuDoc-GAMMA#51]
@thoughtstream Your thoughts on the following are sought. @lizmat @vrurg please extend if need be.

Some time ago I was asking how to implement `DOC use Module`, see [user defined blocks](https://finanalyst.github.io/samples_rakudoc#User-defined_blocks).
At the time there was a strong objection to the idea in that a RakuDoc source should have any knowledge of its environment, or of the output format. I raised the topic again at the Raku Steering Council today. The objection remains.

The DOC phaser does not seem to do anything in Rakudo. So I wonder whether we should remove all mention of DOC phaser from 'The DOC phaser' upto but not including 'If a renderer'.

The specification already provides a mechanism for handling a Custom block that a renderer does not recognise. 
(There is another concern here, but I'm raising another issue for that)

**Comment by thoughtstream** [comment]
If it’s been decided that Raku is not going with the original design
of supporting user-defined _“documentation modules”_ via `DOC` phasers,
then – yes – we should definitely remove any mention of `DOC` from the spec.

**Comment by finanalyst** [comment]
@thoughtstream This is a topic I have wondered about for a while. Sorry that I have not expressed it in this forum.

There is no question about the need for user defined functionality. The question is about signalling that to the renderer.

When developing Raku::Pod::Render and RakuAST::RakuDoc::Render, I could not work out how to use the DOC phasers, or find information about them. So I developed some work arounds.
The approach I have experimented with in the Raku::Pod::Render and Collection distributions is to have an external file with a set of plugins that are loaded and attached to the render-engine object. 
The approach for RakuAST::RakuDoc::Render is to have Plugin classes that attach data to the renderer object when it is instantiated.
In both cases, templates contain the information needed to embed the data in a RakuDoc source in the output format (HTML, md). 

The idea in both is that if a RakuDoc source contains a custom block, then the render engine looks for a template that matches the custom block. If a template is not supplied by a plugin, then it is rendered as unknown.

The problem with `DOC use 'data'` is that it is unclear how to interpret 'data' in a way that does not presume the renderer. All of my customisation approaches make assumptions about the plugins. So templates expect data from the render engine, and return a string-type object that is concatenated into the final string output. 

Currently, a RakuDoc source can be rendered with any renderer that complies with the specification and renderer the rakudociiem-ipsum cource. No presumption is made about the output format or the renderer.

The problem I had with DOC use data is that the semantics of the statement are not clear. When a normal `use` statement is executed, there is a lot of documentation about exported subs, namespaces, roles, etc. In my way of thinking (the software gurus may disagree), I imagine a 'use' statement as a way of adding Raku code into the program I am writing.

How then would a `DOC use` statement differ? By analogy, if a 'use' statement extends a Raku program, would then a 'DOC use' statement extend the RakuDoc source? In which case, we have duplicated the effect with a `=place` block.

**Comment by thoughtstream** [comment]
@finanalyst asked:

> The problem I had with DOC use data is that the semantics of the statement are not clear.
> When a normal `use` statement is executed, there is a lot of documentation about exported
> subs, namespaces, roles, etc. In my way of thinking (the software gurus may disagree), I
> imagine a 'use' statement as a way of adding Raku code into the program I am writing.
>
> How then would a `DOC use` statement differ?

It wouldn’t really differ. It would be a way to add Raku code to the program I am using 
to render my document.

The idea was that a <code>DOC use </code><em><code>Module::Name</code></em> would load code that
enables a (Raku-implemented) renderer to render one or more user-defined
RakuDoc blocks correctly.

The idea was that such _“documentation modules”_ would contain class definitions
for one or more classes, each of which would be able to render an associated
RakuDoc user-defined block. For example, consider:
```rakudoc
DOC use RakuDoc::User::InlineImage;

=begin InlineImage
7283909473612382102386128369123681293689213681268283789012734889
7834978937489732489738479874834623846340237487238947283479382617
6780346732657402187348127387219837819237812973812648294783974982
9507238974823748973289479231749812374892374774839758730487329847
=end   InlineImage
```

The `DOC use RakuDoc::User::InlineImage;` would load the specified module,
which would (presumably) define a `RakuDoc::User::InlineImage` class,
with a suitable `render()` method:
```raku
    class RakuDoc::User::InlineImage {
        method render (RakuDoc::AST::Block $imagedata, RakuDoc::Target $target-format) {
            # Generate and return a suitable rendering of the first argument
            # for an output format specified by the second argument
        }
    }
```

The RakuDoc renderer would be implemented such that, whenever it encountered a user-defined block
it would parse out the block contents into an AST node then execute something like:
```raku
    # Look for a user-defined class to render the corresponding user-defined block...
    with try RakuDoc::User::($USER-DEFINED-BLOCKNAME).?new -> $renderer {
        $rendered-block = $renderer.render($BLOCK-AST, $TARGET-FORMAT);
    }

    # Otherwise, use a loud fallback strategy...
    else {
        warn "Unknown user-defined block type ($USER-DEFINED-BLOCKNAME) encountered";
        $rendered-block = $BLOCK-AST.as-text();
    }
```

In other words, it for any <code>=for </code> <em><code>WhateverBlock</code></em>,
the renderer would look for a <code>RakuDoc::User::</code><em><code>WhateverBlock</code></em> class,
and would then use it to render the block. If no such class were found,
it would fall back to render-as-source-text.

And a `DOC use ...` was to have been the way document writers could extend the renderer
so it knew how to render their user-defined blocks.

**Comment by finanalyst** [comment]
@thoughtstream It is the ***presumably*** in
> which would (presumably) define a RakuDoc::User::InlineImage class
that concerns me. It is at present undefined. So I developed two different approaches, one for `Raku::Pod::Render` and another for `RakuAST::RakuDoc::Render`. 

The 'class' with a 'render' method also demonstrates the inherent problem of this issue.
The `render` method contains a reference to the output format. A RakuDoc source does not 'know' anything about the output format. It is the **renderer**, not the RakuDoc source that defines the output format. This can be seen when the compiler invocation is `raku --rakudoc=HTML somesource.rakudoc`. The class invoked is implicitly `RakuDoc::To::HTML`, which contains a `render` method.

By having a `DOC use` statement in the RakuDoc source, there is an explicit connection to a renderer. 

Another implementation complication that I have encountered is that different output formats require multiple associated assets. So, an HTML output requires CSS (or SCSS) and JS texts that need to be added into the body of the output file. Meanwhile, an epub output requires CSS, together with a manifest file and two table of content files (one for each version of the epub specification).

So when a custom block is to be created (for example to expose the Leaflet Maps service), it requires the loading of JS for the map, and JS for the tile suppliers, and CSS for the maps and tile suppliers. This is all possible, which can be seen in [sample maps](https://finanalyst.github.io/plugins_LeafletMaps).
Moreover, when the HTML is created, the embedded HTML must also set api tokens that are required to access certain high-value tiles from the tile suppliers. In view of this complexity, Leaflet maps cannot easily be incorporated into text or epub output.

As the renderer parses the RakuDoc source, it collects information for the index and table of contents. The data structures for this information is renderer-specific. So, a custom block class needs to know about these structures in order to add information to them, for example in the docs.raku.org site, class typegraphs are added to the docs/Type/xxx.rakudoc files. Ideally, these typegraph images should be referenced in the Table of Contents (they are not using Collection, but they were in Documentable). There is no information at all in the RakuDoc sources about these typegraph images, it was added in by Documentable.

In addition, I found when rendering the docs.raku.org site that the original Documentable renderer created several thousand additional pages which gathered data from the original 700 sources. And the authors of the 700 sources explicitly referenced the "virtual" pages.

Finally, I should add that even when there is a renderer for an output format, it needs to be extensible as well. People will want to be able to change the CSS for the same HTML tags, or use a different sort of tag for some element, eg, use a different table class.

**Comment by finanalyst** [comment]
@thoughtstream thank you for the perl perspective - I was not so involved in the Perl as I am in the Raku community.

Fixing the problems you pointed out in #50 has led down some rabbit holes. I'll fix them, then raise a PR to eliminate `DOC use`

**Comment by thoughtstream** [comment]
Thanks, @finanalyst.

**Comment by finanalyst** [comment]
closing as resolved. idea for `=else` in new issue
